docs/DdnInterShaderDataCacheTracking.md

```diff
@@ -0,0 +1,87 @@
+Detailed Design Note: Tracking inter-shader data exchange for shader caching
+============================================================================
+
+**Status:** draft
+
+LLPC performs full pipeline optimizations of shader pipelines, such as eliminating unused outputs from a shader stage.
+LLPC also enables caching of individual shaders _after_ those full pipeline optimizations have been performed.
+
+When checking each individual shader against the shader cache, the information that affected its compilation but was
+obtained from other shaders must be taken into account. There are essentially two ways of doing so:
+
+  1. Serialize the entire shader into a binary format (LLVM bitcode) and hash the result.
+
+  2. Combine the shader's input hash (and relevant input parts of the pipeline creation info) with a log of all
+     inter-shader data that was used to make decisions during the compilation of the shader so far.
+
+There are various trade-offs between these two approaches. In LLPC, we choose the second one and implement it as
+described in this note.
+
+
+Historical Background
+---------------------
+
+At the time of writing, LLPC already uses variant 2 described in the introduction. It does so by combining various
+pieces of information that have been saved off in a look-aside data structure (`ResourceUsage`) in the
+`PatchCheckShaderCache` pass.
+
+This approach has two downsides:
+
+  1. All this information must be carried in the look-aside data structure at least until this late pass, even though
+     the information may otherwise be obsolete.
+
+  2. Knowledge about what kind of full pipeline optimizations are applied, including some of their details, is
+     centralized in `PatchCheckShaderCache`. This limits the design's extensibility.
+
+
+Implementation overview
+-----------------------
+
+Every function in LLPC gets an attached `!llpc.hash` metadata node.
```

Comment:
Until now, we only build hash for fragment shader and non-fragment shader. it is not very easy to build hash for every stage.
We plan to support partial pipeline compile, in this optimization, we need two kinds of hash
- current shader hash
- current shader hash and ignore the resource node offset in root level.
i.e. we need support multiple hash in !llpc.hash

Reply:
Can you explain this more? What do you mean by ignoring the resource node offset in root level?

Reply:
in partial pipeline compile, we want to ignore resource node offset mismatch in root level between cached pipeline and input pipeline, because we can adjust user data register mapping on final ELF binary for offset mismatch, instead of trig full pipeline compile. you can find the detail in our previous internal discussion "Vulkan DDN for auto-layout pipeline".

Reply:
Ah I see. I still wonder about having two different hashes, though.

Generally, the principle that shader caching is built on is "equal input --> equal output".

Now we're talking about a situation where maybe "different input --> equal output". That sounds like a fragile and risky design. If it is really true, then it should be possible to make the input smaller instead.

So if previously, the situation was "(shader 0, resource node offsets A) --> output" and "(shader 0, resource node offsets B) --> same output", then I would argue that the resource node offsets shouldn't be part of the input to begin with.

This makes conceptual sense if you want to fixup the resource node offsets via ELF relocations in the resulting binary: there is simply no need to feed the resource node offsets into the compiler in that case, since a relocation will be generated anyway.

Reply:
there is simply no need to feed the resource node offsets into the compiler in that case, since a relocation will be generated anyway.
- I am afraid we can't use relocation for resource node offsets, until now, resource node offset mapping is a part of PAL metadata in .note section. I don't think PAL support reloc in this section.

for partial pipeline compile, the situation is
(shader 0, resource node offsets A) --> output A
(shader 0, no offset ) --> output B
we still follow equal input --> equal output, but in second case, we need modify the output B a little as the final output.

Reply:
Wouldn't it then make sense to do the "little modification" of the output in _all_ cases?

Then in all cases it could be "(shader 0, no offset) --> output B", and in all cases the output B would be fixed up at the end.